File: backend_nodejs/src/presentation/handler/updateTodoHandler.ts
import type { NextFunction, Request, Response } from "express";
import z from "zod";
import type { UpdateTodoUsecase } from "@/application_service/usecase/updateTodoUsecase";
import { RepositoryNotFoundError } from "@/domain/repository/errors";
import {
  BadRequestError,
  NotFoundError,
} from "@/presentation/middleware/errorHandler";

const updateTodoRequestSchema = z.object({
  params: z.object({
    id: z.uuid({ version: "v7" }),
  }),
  body: z.object({
    title: z.string().min(2).max(100),
    description: z.string().max(500).optional(),
    completed: z.boolean().optional(),
  }),
});

type UpdateTodoResponse = {
  id: string;
  title: string;
  description: string | null;
  completed: boolean;
};

export class UpdateTodoHandler {
  constructor(private updateTodoUsecase: UpdateTodoUsecase) {}

  handle = async (
    req: Request,
    res: Response<UpdateTodoResponse>,
    next: NextFunction,
  ) => {
    try {
      const parsedReq = updateTodoRequestSchema.parse(req);
      const result = await this.updateTodoUsecase.execute({
        id: parsedReq.params.id,
        title: parsedReq.body.title,
        description: parsedReq.body.description,
      });
      res.json({
        id: result.todo.id,
        title: result.todo.title,
        description: result.todo.description,
        completed: result.todo.isCompleted,
      });
    } catch (error) {
      switch (true) {
        case error instanceof z.ZodError: {
          const message = error.issues
            .map((issue) => `${issue.path}: ${issue.message}`)
            .join(", ");
          return next(
            new BadRequestError(`Invalid request parameters: [${message}]`),
          );
        }
        case error instanceof RepositoryNotFoundError:
          return next(new NotFoundError("Todo not found"));
      }
      return next(error);
    }
  };
}

File: backend_python/src/todo_api/presentation/handler/update_todo_handler.py
from pydantic import BaseModel

from todo_api.application_service.usecase.update_todo_usecase import (
    UpdateTodoUsecase,
    UpdateTodoUsecaseInput,
)
from todo_api.domain.repository.errors import RepositoryNotFoundError
from todo_api.presentation.middleware.error_handler import NotFoundError
from todo_api.utils.uuid import UUID7


class UpdateTodoRequest(BaseModel):
    title: str
    description: str | None = None


class UpdateTodoResponse(BaseModel):
    id: str
    title: str
    description: str | None = None
    completed: bool


class UpdateTodoHandler:
    def __init__(self, update_todo_usecase: UpdateTodoUsecase) -> None:
        self.update_todo_usecase = update_todo_usecase

    def handle(self, id: UUID7, request: UpdateTodoRequest) -> UpdateTodoResponse:
        try:
            result = self.update_todo_usecase.execute(
                UpdateTodoUsecaseInput(id=id, title=request.title, description=request.description)
            )
        except RepositoryNotFoundError:
            raise NotFoundError(message="Todo not found")
        return UpdateTodoResponse(
            id=str(result.todo.id),
            title=result.todo.title,
            description=result.todo.description,
            completed=result.todo.completed,
        )
